File: CS/NewProject/ChartZoomScrollHelper.cs
using DevExpress.XtraCharts;
using System;
using System.Drawing;
using System.Windows.Forms;

namespace NewProject
{
    internal class ChartZoomScrollHelper
    {
        private readonly ChartControl chartControl;
        #region Properties
        public ViewType CurrentViewType { get; set; }
        public Point? FirstPoint { get; private set; }
        public Point? SecondPoint { get; private set; }
        #endregion

        public ChartZoomScrollHelper(ChartControl chart)
        {
            chartControl = chart;
            chartControl.CustomPaint += chartControl1_CustomPaint;
            chartControl.MouseDown += chartControl1_MouseDown;
            chartControl.MouseMove += chartControl1_MouseMove;
            chartControl.MouseUp += chartControl1_MouseUp;
        }
        private void chartControl1_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            FirstPoint = e.Location;
        }

        private void chartControl1_MouseUp(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            if (ShouldZoom(e))
            {
                XYDiagram diagram = (XYDiagram)chartControl.Diagram;

                object argMin, argMax;
                CalcArgVisualRange(diagram, out argMin, out argMax);

                diagram.AxisX.VisualRange.SetMinMaxValues(argMin, argMax);
                FirstPoint = null;
            }

        }

        private void CalcArgVisualRange(XYDiagram diagram, out object argMin, out object argMax)
        {
            int minX = Math.Min(FirstPoint.Value.X, SecondPoint.Value.X);
            int maxX = Math.Max(FirstPoint.Value.X, SecondPoint.Value.X);
            argMin = diagram.PointToDiagram(new Point(minX, SecondPoint.Value.Y)).NumericalArgument;
            argMax = diagram.PointToDiagram(new Point(maxX, SecondPoint.Value.Y)).NumericalArgument;
        }

        private void chartControl1_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            SecondPoint = e.Location;
            if (ShouldScroll())
            {
                XYDiagram diagram = (XYDiagram)chartControl.Diagram;

                double argMin = diagram.PointToDiagram(new Point(FirstPoint.Value.X, SecondPoint.Value.Y)).NumericalArgument;
                double argMax = diagram.PointToDiagram(new Point(SecondPoint.Value.X, SecondPoint.Value.Y)).NumericalArgument;
                double diff = -(argMax - argMin);
                double newMin = (double)diagram.AxisX.VisualRange.MinValue + diff;
                double newMax = (double)diagram.AxisX.VisualRange.MaxValue + diff;
                if ((newMin >= (double)diagram.AxisX.WholeRange.MinValue) && newMax <= (double)diagram.AxisX.WholeRange.MaxValue)
                    diagram.AxisX.VisualRange.SetMinMaxValues(newMin, newMax);
                FirstPoint = SecondPoint;
            }
        }
        private void chartControl1_CustomPaint(object sender, CustomPaintEventArgs e)
        {
            if (ShouldZoom())
                DrawZoomBox(e);
        }

        private void DrawZoomBox(CustomPaintEventArgs e)
        {
            int minX = Math.Min(FirstPoint.Value.X, SecondPoint.Value.X);
            int minY = Math.Min(FirstPoint.Value.Y, SecondPoint.Value.Y);
            int maxX = Math.Max(FirstPoint.Value.X, SecondPoint.Value.X);
            int maxY = Math.Max(FirstPoint.Value.Y, SecondPoint.Value.Y);
            e.Graphics.DrawRectangle(Pens.Black, minX, minY, maxX - minX, maxY - minY);
        }
        private bool ShouldScroll()
        {
            return FirstPoint != null && Control.MouseButtons == MouseButtons.Right;
        }
        private bool ShouldZoom()
        {
            return FirstPoint != null && Control.MouseButtons == MouseButtons.Left;
        }
        private bool ShouldZoom(MouseEventArgs e)
        {
            return FirstPoint != null && e.Button == MouseButtons.Left;
        }
    }
}

File: VB/NewProject/ChartZoomScrollHelper.vb
Imports DevExpress.XtraCharts
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Namespace NewProject
    Friend Class ChartZoomScrollHelper
        Private ReadOnly chartControl As ChartControl
        #Region "Properties"
        Public Property CurrentViewType() As ViewType
        Private privateFirstPoint? As Point
        Public Property FirstPoint() As Point?
            Get
                Return privateFirstPoint
            End Get
            Private Set(ByVal value? As Point)
                privateFirstPoint = value
            End Set
        End Property
        Private privateSecondPoint? As Point
        Public Property SecondPoint() As Point?
            Get
                Return privateSecondPoint
            End Get
            Private Set(ByVal value? As Point)
                privateSecondPoint = value
            End Set
        End Property
        #End Region

        Public Sub New(ByVal chart As ChartControl)
            chartControl = chart
            AddHandler chartControl.CustomPaint, AddressOf chartControl1_CustomPaint
            AddHandler chartControl.MouseDown, AddressOf chartControl1_MouseDown
            AddHandler chartControl.MouseMove, AddressOf chartControl1_MouseMove
            AddHandler chartControl.MouseUp, AddressOf chartControl1_MouseUp
        End Sub
        Private Sub chartControl1_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
            FirstPoint = e.Location
        End Sub

        Private Sub chartControl1_MouseUp(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
            If ShouldZoom(e) Then
                Dim diagram As XYDiagram = CType(chartControl.Diagram, XYDiagram)

                Dim argMin As Object = Nothing, argMax As Object = Nothing
                CalcArgVisualRange(diagram, argMin, argMax)

                diagram.AxisX.VisualRange.SetMinMaxValues(argMin, argMax)
                FirstPoint = Nothing
            End If

        End Sub

        Private Sub CalcArgVisualRange(ByVal diagram As XYDiagram, ByRef argMin As Object, ByRef argMax As Object)
            Dim minX As Integer = Math.Min(FirstPoint.Value.X, SecondPoint.Value.X)
            Dim maxX As Integer = Math.Max(FirstPoint.Value.X, SecondPoint.Value.X)
            argMin = diagram.PointToDiagram(New Point(minX, SecondPoint.Value.Y)).NumericalArgument
            argMax = diagram.PointToDiagram(New Point(maxX, SecondPoint.Value.Y)).NumericalArgument
        End Sub

        Private Sub chartControl1_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs)
            SecondPoint = e.Location
            If ShouldScroll() Then
                Dim diagram As XYDiagram = CType(chartControl.Diagram, XYDiagram)

                Dim argMin As Double = diagram.PointToDiagram(New Point(FirstPoint.Value.X, SecondPoint.Value.Y)).NumericalArgument
                Dim argMax As Double = diagram.PointToDiagram(New Point(SecondPoint.Value.X, SecondPoint.Value.Y)).NumericalArgument
                Dim diff As Double = -(argMax - argMin)
                Dim newMin As Double = CDbl(diagram.AxisX.VisualRange.MinValue) + diff
                Dim newMax As Double = CDbl(diagram.AxisX.VisualRange.MaxValue) + diff
                If (newMin >= CDbl(diagram.AxisX.WholeRange.MinValue)) AndAlso newMax <= CDbl(diagram.AxisX.WholeRange.MaxValue) Then
                    diagram.AxisX.VisualRange.SetMinMaxValues(newMin, newMax)
                End If
                FirstPoint = SecondPoint
            End If
        End Sub
        Private Sub chartControl1_CustomPaint(ByVal sender As Object, ByVal e As CustomPaintEventArgs)
            If ShouldZoom() Then
                DrawZoomBox(e)
            End If
        End Sub

        Private Sub DrawZoomBox(ByVal e As CustomPaintEventArgs)
            Dim minX As Integer = Math.Min(FirstPoint.Value.X, SecondPoint.Value.X)
            Dim minY As Integer = Math.Min(FirstPoint.Value.Y, SecondPoint.Value.Y)
            Dim maxX As Integer = Math.Max(FirstPoint.Value.X, SecondPoint.Value.X)
            Dim maxY As Integer = Math.Max(FirstPoint.Value.Y, SecondPoint.Value.Y)
            e.Graphics.DrawRectangle(Pens.Black, minX, minY, maxX - minX, maxY - minY)
        End Sub
        Private Function ShouldScroll() As Boolean
            Return FirstPoint IsNot Nothing AndAlso Control.MouseButtons = MouseButtons.Right
        End Function
        Private Function ShouldZoom() As Boolean
            Return FirstPoint IsNot Nothing AndAlso Control.MouseButtons = MouseButtons.Left
        End Function
        Private Function ShouldZoom(ByVal e As MouseEventArgs) As Boolean
            Return FirstPoint IsNot Nothing AndAlso e.Button = MouseButtons.Left
        End Function
    End Class
End Namespace